### ISSUE TITLE
TypeError: superclass mismatch for class OCI8

### ISSUE THREAD

**Issue by yahonda** [kubo/ruby-oci8#80]
Using the latest ruby-oci8, oracle enhanced unit tests gets TypeError: superclass mismatch for class OCI8.

```ruby
$ rake spec
==> Loading config from ENV or use default
==> Running specs with MRI version 2.2.2
rake aborted!
TypeError: superclass mismatch for class OCI8
/home/yahonda/git/ruby-oci8/lib/oci8.rb:70:in `require'
/home/yahonda/git/ruby-oci8/lib/oci8.rb:70:in `<top (required)>'
/home/yahonda/git/oracle-enhanced/spec/spec_helper.rb:20:in `require'
/home/yahonda/git/oracle-enhanced/spec/spec_helper.rb:20:in `<top (required)>'
/home/yahonda/git/oracle-enhanced/Rakefile:39:in `require'
/home/yahonda/git/oracle-enhanced/Rakefile:39:in `block in <top (required)>'
Tasks: TOP => spec => clear
(See full trace by running task with --trace)
$
```

- steps to reproduce

```ruby
$ ruby -v
ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-linux]
$ bundle -v
Bundler version 1.10.4
$ bundle config
Settings are listed in order of priority. The top value will be used.
local.activerecord-oracle_enhanced-adapter
Set for the current user (/home/yahonda/.bundle/config): "/home/yahonda/git/oracle-enhanced"

disable_local_branch_check
Set for the current user (/home/yahonda/.bundle/config): "true"

local.activerecord
Set for the current user (/home/yahonda/.bundle/config): "/home/yahonda/git/rails"

loral.arel
Set for the current user (/home/yahonda/.bundle/config): "/home/yahonda/git/arel"

local.ruby-oci8
Set for the current user (/home/yahonda/.bundle/config): "/home/yahonda/git/ruby-oci8"

gem.mit
Set for the current user (/home/yahonda/.bundle/config): "0"

gem.test
Set for the current user (/home/yahonda/.bundle/config): "rspec"

gem.coc
Set for the current user (/home/yahonda/.bundle/config): "0"

$ git clone git@github.com:rsim/oracle-enhanced.git
$ cd oracle-enhanced
$ bundle
Resolving dependencies...
Using rake 10.4.2
Using i18n 0.7.0
Using json 1.8.3
Using minitest 5.7.0
Using thread_safe 0.3.5
Using tzinfo 1.2.2
Using activesupport 4.2.2 from git://github.com/rails/rails.git (at /home/yahonda/git/rails)
Using builder 3.2.2
Using erubis 2.7.0
Using mini_portile 0.6.2
Using nokogiri 1.6.6.2
Using rails-deprecated_sanitizer 1.0.3
Using rails-dom-testing 1.0.6
Using loofah 2.0.2
Using rails-html-sanitizer 1.0.2
Using actionview 4.2.2 from git://github.com/rails/rails.git (at /home/yahonda/git/rails)
Using rack 1.6.4
Using rack-test 0.6.3
Using actionpack 4.2.2 from git://github.com/rails/rails.git (at /home/yahonda/git/rails)
Using activemodel 4.2.2 from git://github.com/rails/rails.git (at /home/yahonda/git/rails)
Using arel 6.0.0 from git://github.com/rails/arel.git (at 6-0-stable)
Using activerecord 4.2.2 from git://github.com/rails/rails.git (at /home/yahonda/git/rails)
Using activerecord-deprecated_finders 1.0.4
Using addressable 2.3.8
Using bundler 1.10.4
Using descendants_tracker 0.0.4
Using diff-lcs 1.2.5
Using multipart-post 2.0.0
Using faraday 0.9.1
Using git 1.2.9.1
Using hashie 3.4.2
Using multi_json 1.11.1
Using jwt 1.5.0
Using multi_xml 0.5.5
Using oauth2 1.0.0
Using github_api 0.12.3
Using highline 1.7.2
Using rdoc 4.2.0
Using jeweler 2.0.1
Using journey 2.0.0.20120723141804 from git://github.com/rails/journey.git (at master)
Using thor 0.19.1
Using railties 4.2.2 from git://github.com/rails/rails.git (at /home/yahonda/git/rails)
Using rspec-core 2.99.2
Using rspec-expectations 2.99.2
Using rspec-mocks 2.99.4
Using rspec 2.99.0
Using ruby-oci8 2.2.0 from git://github.com/kubo/ruby-oci8.git (at /home/yahonda/git/ruby-oci8)
Using ruby-plsql 0.5.3
Bundle complete! 13 Gemfile dependencies, 48 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.
$ rake spec
==> Loading config from ENV or use default
==> Running specs with MRI version 2.2.2
rake aborted!
TypeError: superclass mismatch for class OCI8
/home/yahonda/git/ruby-oci8/lib/oci8.rb:70:in `require'
/home/yahonda/git/ruby-oci8/lib/oci8.rb:70:in `<top (required)>'
/home/yahonda/git/oracle-enhanced/spec/spec_helper.rb:20:in `require'
/home/yahonda/git/oracle-enhanced/spec/spec_helper.rb:20:in `<top (required)>'
/home/yahonda/git/oracle-enhanced/Rakefile:39:in `require'
/home/yahonda/git/oracle-enhanced/Rakefile:39:in `block in <top (required)>'
Tasks: TOP => spec => clear
(See full trace by running task with --trace)
$
```

- With trace 

```ruby
$ rake spec --trace
** Invoke spec (first_time)
** Invoke clear (first_time)
** Execute clear
==> Loading config from ENV or use default
==> Running specs with MRI version 2.2.2
rake aborted!
TypeError: superclass mismatch for class OCI8
/home/yahonda/git/ruby-oci8/lib/oci8.rb:70:in `require'
/home/yahonda/git/ruby-oci8/lib/oci8.rb:70:in `<top (required)>'
/home/yahonda/git/oracle-enhanced/spec/spec_helper.rb:20:in `require'
/home/yahonda/git/oracle-enhanced/spec/spec_helper.rb:20:in `<top (required)>'
/home/yahonda/git/oracle-enhanced/Rakefile:39:in `require'
/home/yahonda/git/oracle-enhanced/Rakefile:39:in `block in <top (required)>'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:240:in `call'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:240:in `block in execute'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:235:in `each'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:235:in `execute'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:179:in `block in invoke_with_call_chain'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:172:in `invoke_with_call_chain'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:201:in `block in invoke_prerequisites'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:199:in `each'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:199:in `invoke_prerequisites'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:178:in `block in invoke_with_call_chain'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:172:in `invoke_with_call_chain'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:165:in `invoke'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:150:in `invoke_task'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:106:in `block (2 levels) in top_level'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:106:in `each'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:106:in `block in top_level'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:115:in `run_with_threads'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:100:in `top_level'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:78:in `block in run'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:176:in `standard_exception_handling'
/home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:75:in `run'
/home/yahonda/.rvm/rubies/ruby-2.2.2/bin/rake:33:in `<main>'
Tasks: TOP => spec => clear
$
```

**Comment by kubo** [comment]
Could you post `bundle show`?
I cannot reproduce the issue by the steps because of the following dependency error.
```shell
$ git clone git@github.com:rsim/oracle-enhanced.git
Cloning into 'oracle-enhanced'...
remote: Counting objects: 5991, done.
remote: Compressing objects: 100% (12/12), done.
remote: Total 5991 (delta 3), reused 0 (delta 0), pack-reused 5977
Receiving objects: 100% (5991/5991), 4.24 MiB | 552 KiB/s, done.
Resolving deltas: 100% (2709/2709), done.
$ cd oracle-enhanced/
$ bundle config
Settings are listed in order of priority. The top value will be used.
$ bundle
Fetching git://github.com/rails/rails.git
Fetching git://github.com/rails/arel.git
Fetching git://github.com/rails/journey.git
Fetching git://github.com/kubo/ruby-oci8.git
Fetching gem metadata from http://rubygems.org/.........
Fetching version metadata from http://rubygems.org/...
Fetching dependency metadata from http://rubygems.org/..
Resolving dependencies...
Bundler could not find compatible versions for gem "arel":
  In Gemfile:
    activerecord (>= 0) ruby depends on
      arel (= 7.0.0.alpha) ruby

    arel (>= 0) ruby
Could not find gem 'arel (= 7.0.0.alpha) ruby', which is required by gem 'activerecord (>= 0) ruby', in any of the sources.
$ ruby -v
ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-linux]
$ gem list

*** LOCAL GEMS ***

bigdecimal (1.2.6)
bundler (1.10.4)
io-console (0.4.3)
json (1.8.1)
minitest (5.4.3)
power_assert (0.2.2)
psych (2.0.8)
rake (10.4.2)
rake-compiler (0.9.5)
rdoc (4.2.0)
test-unit (3.0.8)
```
I suppose that somebody requires `oci8/version.rb` before `oci8lib_xxx.so`.

**Comment by yahonda** [comment]
Here is the output of `bundle show`

```ruby
2.2.2@rails42 [ master ~/git/oracle-enhanced]$ bundle show
Gems included by the bundle:
  * actionpack (4.2.2 24637e5)
  * actionview (4.2.2 24637e5)
  * activemodel (4.2.2 24637e5)
  * activerecord (4.2.2 24637e5)
  * activerecord-deprecated_finders (1.0.4)
  * activesupport (4.2.2 24637e5)
  * addressable (2.3.8)
  * arel (6.0.0 fdfe66a)
  * builder (3.2.2)
  * bundler (1.10.4)
  * descendants_tracker (0.0.4)
  * diff-lcs (1.2.5)
  * erubis (2.7.0)
  * faraday (0.9.1)
  * git (1.2.9.1)
  * github_api (0.12.3)
  * hashie (3.4.2)
  * highline (1.7.2)
  * i18n (0.7.0)
  * jeweler (2.0.1)
  * journey (2.0.0.20120723141804 965b287)
  * json (1.8.3)
  * jwt (1.5.1)
  * loofah (2.0.2)
  * mini_portile (0.6.2)
  * minitest (5.7.0)
  * multi_json (1.11.1)
  * multi_xml (0.5.5)
  * multipart-post (2.0.0)
  * nokogiri (1.6.6.2)
  * oauth2 (1.0.0)
  * rack (1.6.4)
  * rack-test (0.6.3)
  * rails-deprecated_sanitizer (1.0.3)
  * rails-dom-testing (1.0.6)
  * rails-html-sanitizer (1.0.2)
  * railties (4.2.2 24637e5)
  * rake (10.4.2)
  * rdoc (4.2.0)
  * rspec (2.99.0)
  * rspec-core (2.99.2)
  * rspec-expectations (2.99.2)
  * rspec-mocks (2.99.4)
  * ruby-oci8 (2.2.0 9e6a11a)
  * ruby-plsql (0.5.3)
  * thor (0.19.1)
  * thread_safe (0.3.5)
  * tzinfo (1.2.2)
$
```

**Comment by yahonda** [comment]
I use local rails and arel as follows. replace /home/yahonda/git to your local directory and
perform these git clone to use local git repository.

```
$ git clone git@github.com:rails/rails.git
$ cd rails
$ git checkout 4-2-stable
```

```

$ git clone git@github.com:rails/arel.git
$ cd arel/
$ git checkout 6-0-stable
```


```ruby
$ bundle config
Settings are listed in order of priority. The top value will be used.
local.activerecord-oracle_enhanced-adapter
Set for the current user (/home/yahonda/.bundle/config): "/home/yahonda/git/oracle-enhanced"

disable_local_branch_check
Set for the current user (/home/yahonda/.bundle/config): "true"

local.activerecord
Set for the current user (/home/yahonda/.bundle/config): "/home/yahonda/git/rails"

loral.arel
Set for the current user (/home/yahonda/.bundle/config): "/home/yahonda/git/arel"

local.ruby-oci8
Set for the current user (/home/yahonda/.bundle/config): "/home/yahonda/git/ruby-oci8"

gem.mit
Set for the current user (/home/yahonda/.bundle/config): "0"

gem.test
Set for the current user (/home/yahonda/.bundle/config): "rspec"

gem.coc
Set for the current user (/home/yahonda/.bundle/config): "0"

$
```

**Comment by kubo** [comment]
Thank you!
Fixed by 9a0af41f457d77fa1e380f04ef71e94987afb789.
I reproduced the issue and found that `rake spec` loaded `ruby-oci8.gemspec`, which  included `oci8/version.rb`, before `require "oci8"`.

**Comment by yahonda** [comment]
Thanks for the fix. I also verified it has fixed in master branch.
